### ISSUE TITLE
MdDialog beforeClosed() feature

### ISSUE THREAD

**Issue by VagrantAI-c** [angular/material2#5931]
#### Bug, feature request, or proposal:
Feature request

#### What is the expected behavior?
If there is an `afterClosed()` event hook in dialogRef, why isn't there is an `beforeClosed()` hook?

#### What is the current behavior?
MdDialog currently supports only `afterClosed()` hook

#### What is the use-case or motivation for changing an existing behavior?
In my app dialogs has its own router urls. Since i can only detect dialog closing only after animation has been passed, opening a new dialog while the older one is still closing might produce router errors.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
Angular 4, Angular Material 2.0.0-beta.8

**Comment by julianobrasil** [comment]
@VagrantAI-c, you can encapsulate your dialog calling inside a component to do it (take a look at this [stackoverflow.com anwer](https://stackoverflow.com/a/39966683/6433166)).

But if you follow the template issue, there's a section where you can show an use case that justify this implementation (I think it would be nice to have this feature).

Not following the issue template can lead the team to close your issue. Modify it as a feature request.

**Comment by willshowell** [comment]
Duplicate of https://github.com/angular/material2/issues/4647

**Comment by jelbourn** [comment]
Dupe of #4647
